### PR TITLE
CXP-1255: Product value filter - locale

### PR DESCRIPTION
**Pull request by guillaumefer-akeneo**
**Description (for Contributor and Core Developer)**

https://user-images.githubusercontent.com/93598018/189151945-e94db5ca-1471-43b9-87f7-08fe56577862.mp4
https://user-images.githubusercontent.com/93598018/189151975-42128610-e09c-4e69-a81e-0cf9326efb84.mp4

**Definition Of Done (for Core Developer only)**

- [ ] Tests
- [ ] Migration & Installer
- [ ] PM Validation (Story)
- [ ] Changelog (maintenance bug fixes)
- [ ] Tech Doc

### CODE REVIEW

File: components/catalogs/back/src/Application/Persistence/GetLocalesQueryInterface.php [hunk 1]
```diff
@@ -13,5 +13,5 @@ interface GetLocalesQueryInterface
     /**
      * @return array<array-key, array{code: string, label: string}>
      */
-    public function execute(): array;
+    public function execute(int $page = 1, int $limit = 20): array;
```
Comment (masacc) [inline components/catalogs/back/src/Application/Persistence/GetLocalesQueryInterface.php:16]: I think there is no need to paginate this query as there will never be much more than 200 locales in the worst case. That's why there is a simple hook in front `useLocale`. It simplifies the code and makes fewer HTTP calls.

File: components/catalogs/back/src/Application/Persistence/GetLocalesByCodeQueryInterface.php [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+interface GetLocalesByCodeQueryInterface
+{
+    /**
+     * @param array<string> $codes
+     *
+     * @return array<array-key, array{code: string, label: string}>
+     */
+    public function execute(array $codes, int $page = 1, int $limit = 20): array;
```
Comment (masacc) [inline components/catalogs/back/src/Application/Persistence/GetLocalesByCodeQueryInterface.php:18]: same here, I don't think we need to paginate

File: components/catalogs/back/src/Infrastructure/Validation/ProductValueFilters/FilterContainsActivatedLocaleValidator.php [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+        }
+
+        if (!\is_string($value) || '' === $value) {
+            return;
+        }
+
+        /** @var array<LocaleInterface> $locales< */
+        $locales = $this->getLocalesByCodeQuery->execute([$value]);
+
+        if (!\count($locales)) {
```
Comment (masacc) [inline components/catalogs/back/src/Infrastructure/Validation/ProductValueFilters/FilterContainsActivatedLocaleValidator.php:39]:
```suggestion
        if (0 === \count($locales)) {
```
or I think rector will propose you
```suggestion
        if ([] === $locales) {
```

File: components/catalogs/back/tests/Integration/Infrastructure/Controller/Internal/GetLocalesActionTest.php [hunk 4]
```diff
@@ -48,4 +48,38 @@ public function testItGetsLocales(): void
...
+    public function testItGetsLocalesByCode(): void
+    {
+        // Locales are only activated when used in a channel
+        $this->createChannel('mobile', ['fr_FR', 'en_US']);
+
+        $client = $this->getAuthenticatedInternalApiClient();
+
+        $client->request(
+            'GET',
+            '/rest/catalogs/locales?codes=en_US,fr_FR',
```
Comment (masacc) [inline components/catalogs/back/tests/Integration/Infrastructure/Controller/Internal/GetLocalesActionTest.php:61]: Can you test with only one code or like this but with a third one activated to be sure it really filters by codes?

File: components/catalogs/back/tests/Integration/Infrastructure/Controller/Internal/GetLocalesActionTest.php [hunk 5]
```diff
@@ -48,4 +48,38 @@ public function testItGetsLocales(): void
...
+                'label' => 'English (United States)',
+            ],
+            [
+                'code' => 'fr_FR',
+                'label' => 'French (France)',
+            ],
+        ];
+        Assert::assertEquals($expectedLocales, $locales);
+    }
 }
```
Comment (masacc) [inline components/catalogs/back/tests/Integration/Infrastructure/Controller/Internal/GetLocalesActionTest.php:85]: You forgot some unit test for this `Action`

File: components/catalogs/front/src/components/ProductSelection/components/AnyLocaleInput.tsx [hunk 6]
```diff
@@ -26,7 +26,7 @@ const AnyLocaleInput: FC<Props> = ({state, onChange, isInvalid}) => {
             onChange={v => onChange({...state, locale: v})}
             clearable={false}
             invalid={isInvalid}
-            placeholder={translate('akeneo_catalogs.product_selection.locale.label')}
+            placeholder={translate('akeneo_catalogs.product_selection.filters.locale.label')}
```
Comment (masacc) [inline components/catalogs/front/src/components/ProductSelection/components/AnyLocaleInput.tsx:29]:
```suggestion
            placeholder={translate('akeneo_catalogs.product_selection.locale.label')}
```

this is not filters, this is input for product selection

File: components/catalogs/front/src/components/ProductSelection/components/ScopableLocaleInput.tsx [hunk 7]
```diff
@@ -26,7 +26,7 @@ const ScopableLocaleInput: FC<Props> = ({state, onChange, isInvalid}) => {
             onChange={v => onChange({...state, locale: v})}
             clearable={false}
             invalid={isInvalid}
-            placeholder={translate('akeneo_catalogs.product_selection.locale.label')}
+            placeholder={translate('akeneo_catalogs.product_selection.filters.locale.label')}
```
Comment (masacc) [inline components/catalogs/front/src/components/ProductSelection/components/ScopableLocaleInput.tsx:29]:
```suggestion
            placeholder={translate('akeneo_catalogs.product_selection.locale.label')}
```

File: components/catalogs/front/src/components/ProductSelection/criteria/CompletenessCriterion/CompletenessLocaleInput.tsx [hunk 8]
```diff
@@ -22,7 +22,7 @@ const CompletenessLocaleInput: FC<Props> = ({state, onChange, isInvalid}) => {
             onChange={v => onChange({...state, locale: v})}
             clearable={false}
             invalid={isInvalid}
-            placeholder={translate('akeneo_catalogs.product_selection.locale.label')}
+            placeholder={translate('akeneo_catalogs.product_selection.filters.locale.label')}
```
Comment (masacc) [inline components/catalogs/front/src/components/ProductSelection/criteria/CompletenessCriterion/CompletenessLocaleInput.tsx:25]:
```suggestion
            placeholder={translate('akeneo_catalogs.product_selection.locale.label')}
```

**Review by willy-ahva — APPROVED** [review]

File: components/catalogs/front/src/components/ProductValueFilters/hooks/useLocalesByCodes.ts [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+            headers: {
+                'X-Requested-With': 'XMLHttpRequest',
+            },
+        });
+
+        const locales: Locale[] = await response.json();
+
+        const localeCodes = locales.map(locale => locale.code);
+
+        const desactivatedLocaleCodes = codes.filter(code => !localeCodes.includes(code));
```
Comment (masacc) [inline components/catalogs/front/src/components/ProductValueFilters/hooks/useLocalesByCodes.ts:26]:
```suggestion
        const deactivatedLocaleCodes = codes.filter(code => !localeCodes.includes(code));
```

File: components/catalogs/front/src/components/ProductValueFilters/hooks/useLocalesByCodes.unit.ts [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+            {
+                code: 'en_US',
+                label: 'English (United States)',
+            },
+            {
+                code: 'fr_FR',
+                label: 'French (France)',
+            },
+        ])
+    );
```
Comment (masacc) [inline components/catalogs/front/src/components/ProductValueFilters/hooks/useLocalesByCodes.unit.ts:22]: Rather use : 
```
beforeEach(() => {
    mockFetchResponses([
        {
            url: '/rest/catalogs/locales?codes=en_US,fr_FR,de_DE,
            json: [EN, FR, DE],
        },
    ]);
});
```

like in other recent tests

File: components/catalogs/front/src/components/ProductValueFilters/hooks/useLocalesByCodes.unit.ts [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+                label: 'English (United States)',
+            },
+            {
+                code: 'fr_FR',
+                label: 'French (France)',
+            },
+            {
+                code: 'de_DE',
+                label: '[de_DE]',
+            },
```
Comment (masacc) [inline components/catalogs/front/src/components/ProductValueFilters/hooks/useLocalesByCodes.unit.ts:53]: Your mock only returns 2 locales and you expect 3 here.

Comment (guillaumefer-akeneo) [reply components/catalogs/front/src/components/ProductValueFilters/hooks/useLocalesByCodes.unit.ts:53]: This is the expected result.
You can see in the URL `?codes=en_US,fr_FR,de_DE`, it happens when a locale has been deactivated so the backend returns a response with only 2 locales.
The useLocalesByCodes adds missing locales with a special label wrapped with brackets `[de_DE]`, when you save, a violation is displayed.
I add a comment.

File: components/catalogs/back/src/Application/Persistence/GetLocalesByCodeQueryInterface.php (hunk 2 above)
Comment (guillaumefer-akeneo) [reply components/catalogs/back/src/Application/Persistence/GetLocalesByCodeQueryInterface.php:18]: We had talked with the squad about copying the infiniteScroll to save time and simplify maintenance.
Anyway, I made a 2nd PR (based on this PR) to remove the pagination : https://github.com/akeneo/pim-community-dev/pull/17890

File: components/catalogs/back/tests/Unit/Infrastructure/Controller/Internal/GetLocalesActionTest.php [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+        return [
+            'limit and page are positive' => [
+                [
+                    'page' => 1,
+                    'limit' => 1,
+                    'codes' =>'en_US,fr_FR',
+                ],
+            ],
+        ];
+    }
```
Comment (masacc) [inline components/catalogs/back/tests/Unit/Infrastructure/Controller/Internal/GetLocalesActionTest.php:153]: Is it really useful to have a dataProvider for only one case?

**Pull request merged**